### PR TITLE
Receipt fixes

### PR DESCRIPTION
**Pull request by benjamincburns**
Makes it so that receipts are returned when EVM runtime errors occur (e.g. `REVERT`). Adds missing `status`, `from`, and `to` fields to the transaction receipt.

### CODE REVIEW

**Comment by aludvik** [comment]
@benjamincburns Since this is a pretty substantial change (some previous transactions that were previously invalid are now valid), can you summarize the discussion from RocketChat about why this is the desired behavior, what the pros and cons are, and what use cases this support.

**Comment by benjamincburns** [comment]
Go formatting issues are now fixed.

**Comment by benjamincburns** [comment]
@aludvik sure thing. 

To be clear, this change doesn't allow for contract state updates when transaction execution fails. It simply records the transaction execution failure in a transaction receipt which is appropriate for the transaction family.

Prior to these changes there was no way for a client which submitted its transactions via Seth's RPC interface to determine whether the transaction was executed, and what the result of that execution was.

Via the Ethereum RPC interface, the usual transaction flow looks like this:

1. Client submits a message call transaction to a contract address via either the `eth_sendTransaction` or the `eth_sendRawTransaction` RPC call
2. Server responds with RPC response containing transaction hash in the `result` field
3. Client uses the transaction hash returned by the server to poll for the result of the transaction via `eth_getTransactionReceipt`.
4. Until the transaction is mined into a block, the server always returns `null` in response to the client's `eth_getTransactionReceipt` request.
5. When the transaction is mined into a block, regardless of whether the transaction succeeded or failed,  the server responds to `eth_getTransactionReceipt` with a valid transaction receipt.
6. The client inspects the `status` field in `eth_getTransactionReceipt` to determine whether their transaction succeeded or failed.

If no receipt is produced when a message call to a contract address fails due to a runtime error (e.g. due to the `REVERT` opcode), then the client can't tell the difference between transaction failure, and the transaction never starting.

**Comment by benjamincburns** [comment]
As to what use-cases this supports, I'm working on integrating the Truffle toolsuite with Sawtooth-Seth. See the blocked tickets on the [commit-referenced Jira issue](https://jira.hyperledger.org/browse/STL-1460) for more info.

**Comment by benjamincburns** [comment]
Looks like I had some rust formatting issues as well - those should be fixed up now.

File: rpc/tests/test_seth_rpc.py [hunk 1]
```diff
@@ -238,6 +239,9 @@ def test_get_block_by_hash(self):
         self.assertEqual(result["stateRoot"], "0x" + self.state_root)
         self.assertEqual(result["gasUsed"], hex(self.gas))
         self.assertEqual(result["transactions"][0], "0x" + self.txn_id)
+        self.assertEqual(result["status"], "0x" + self.status)
```
Comment (knkski) [inline rpc/tests/test_seth_rpc.py]: Looks like this line is causing issues in Jenkins atm:

https://build.sawtooth.me/job/Sawtooth-Hyperledger/job/sawtooth-seth/view/change-requests/job/PR-77/7/console

File: rpc/tests/test_seth_rpc.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ def setUpClass(cls):
         # account values
         cls.public_key = "036d7bb6ca0fd581eb037e91042320af97508003264f08545a9db134df215f373e"
         cls.account_address = "434d46456b6973a678b77382fca0252629f4389f"
+        cls.account_address_b = bytes([0x43, 0x4d, 0x46, 0x45, 0x6b, 0x69, 0x73, 0xa6, 0x78, 0xb7, 0x73, 0x82, 0xfc, 0xa0, 0x25, 0x26, 0x29, 0xf4, 0x38, 0x9f])
```
Comment (knkski) [inline rpc/tests/test_seth_rpc.py]: Might be worth writing this as `cls.account_address_b = bytes.fromhex(cls.account_address)` so that the two definitions don't get out of sync

File: rpc/tests/test_seth_rpc.py [hunk 3]
```diff
@@ -1065,15 +1068,17 @@ def test_get_log_filter_changes(self):
...
         ))
 
+        receipt = SethTransactionReceipt(
+                gas_used=self.gas,
+                return_value=self.return_value_b,
+                contract_address=self.contract_address_b,
+                status=self.status,
+                to=self.contract_address_b,
+                )
+        setattr(receipt, 'from', self.account_address_b)
```
Comment (knkski) [inline rpc/tests/test_seth_rpc.py]: Wondering what the purpose of `setattr` is here vs just passing it into the constructor

File: rpc/tests/test_seth_rpc.py (hunk 2 above)
Comment (benjamincburns) [reply rpc/tests/test_seth_rpc.py]: good point - I missed that helper function - will fix.

File: rpc/tests/test_seth_rpc.py (hunk 3 above)
Comment (benjamincburns) [reply rpc/tests/test_seth_rpc.py]: Sadly due to a limitation of python you can't use reserved words as keyword arguments for a function. The protobuf devs [seem to be aware of that limitation, but it's as yet unresolved](https://github.com/protocolbuffers/protobuf/blob/master/python/google/protobuf/internal/python_message.py#L220-L236), and `setattr` is the only viable workaround I could find.

Comment (benjamincburns) [reply rpc/tests/test_seth_rpc.py]: I'll add a comment to make this more clear, since I'm already going to do that `bytes.fromhex` fix you mentioned elsewhere...

Comment (knkski) [reply rpc/tests/test_seth_rpc.py]: Cool, makes sense. Another option is that you could do is `SethTransactionReceipt(**{'from': self.account_address_b, ...})`, but either way works

**Comment by aludvik** [comment]
@benjamincburns Thanks for the response. Am I correct in understanding there are three types of validity with Ethereum transactions? 
1. Totally valid - The transaction "object" was valid and the execution was a success
2. Valid but failed -  The transaction "object" was valid but the execution failed
3. Invalid - The transaction "object" was invalid because of something like a bad signature, bad payload, incorrect nonce.

Is that characterization accurate? Does the nonce still get updated if the transaction execution fails?

**Comment by benjamincburns** [comment]
@aludvik sorry that I didn't see your question above! Yes, your characterization is accurate. The nonce gets updated in cases 1 and 2, as in these cases a transaction receipt is generated. Transaction receipts are included in block hash calculations, and as such they're stored on chain under the transaction receipt storage trie data structure.

**Comment by benjamincburns** [comment]
Noticed that this branch has gone a bit stale, so I fixed up the outstanding conflicts. I've also ruthlessly culled the projects/issues I'm watching, so with any luck I won't miss future notifications on this PR!

**Comment by benjamincburns** [comment]
> FAIL: Missing Signed-Off Field
I suppose that's what I get for fixing conflicts with GitHub's internal tooling - will fix and force-push.

**Comment by vaporos** [comment]
@benjamincburns Looks great.

Can you rebase on top of master (instead of adding the merge commit) and force-push?

Also, in your commit messages, can you remove the STL- from the first line? When we reference bugs (which isn't often currently), we prefer to follow https://chris.beams.io/posts/git-commit/ which is to put "Resolves: xxx" at the bottom of the commit message.

**Comment by benjamincburns** [comment]
I'm not sure where this left off and it's been quite some time since the original submission. I've moved on to other projects in that time, so I think it's best if I go ahead and close.
